src/metatracer.rs

```diff
@@ -54,68 +94,131 @@ const PHASE_COUNTING: u32 = 0b10 << 30; // The value specifies the current hot c
         // we simply retry the whole operation.
+        let mut trace: Option<Box<Trace>> = None;
         loop {
             let pack = &loc.pack;
-            // We need Acquire ordering, as PHASE_COMPILED will need to read information written to
-            // external data as a result of the PHASE_TRACING -> PHASE_COMPILED transition.
             let lp = pack.load(Ordering::Acquire);
             match lp & PHASE_TAG {
                 PHASE_COUNTING => {
                     let count = lp & !PHASE_TAG;
-                    let new_pack;
                     if count >= self.hot_threshold {
-                        new_pack = PHASE_TRACING;
+                        if pack.compare_and_swap(lp, PHASE_TRACING, Ordering::Release) == lp {
+                            // Only start tracing once updating the pack has succeeded to avoid
+                            // tracing multiple iterations of the phase update loop.
+                            let mut res = Ok(());
+                            with_thread_tracer(&self.tracer, |thr_tracer| res = thr_tracer.start_tracing());
+                            match res {
+                                Ok(_) => break,
+                                Err(HWTracerError::TracerState(TracerState::Started)) => {
+                                    // This thread's tracer is already tracing something (it even
+                                    // could have been initiated from another instance of
+                                    // MetaTracer in the same thread). Roll back the pack to as it
+                                    // was before. We can't clash with another thread at this
+                                    // point, as no one else could have mutated this location since
+                                    // we tagged it PHASE_TRACING.
+                                    pack.store(PHASE_COUNTING | count, Ordering::Relaxed);
+                                    break;
+                                },
+                                Err(e) => unreachable!("{}", e),
+                            }
+                        }
                     } else {
-                        new_pack = PHASE_COUNTING | (count + 1);
-                    }
-                    if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
-                        break;
+                        let new_pack = PHASE_COUNTING | (count + 1);
+                        if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
+                            break;
+                        }
                     }
                 },
                 PHASE_TRACING => {
-                    if pack.compare_and_swap(lp, PHASE_COMPILED, Ordering::Release) == lp {
+                    if trace.is_none() {
+                        // It wouldn't make sense to stop tracing multiple times, so stash the
```

Comment:
I have no idea what this comment means.

Reply:
If we couldn't update the pack, then we don't want to stop the tracer a second time around the loop. Admittedly, I find this fiddly too.

Reply:
Ah, can we rephrase then? The comment really isn't clear at the moment.